internal/exec/stages/disks/luks.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	for _, luks := range config.Storage.Luks {
+		// track whether Ignition creates the KeyFile
+		// so that it can be removed creation
+		var ignitionCreatedKeyFile bool
+		// create keyfile inside of tmpfs, it will be copied to the
+		// sysroot by ignition-copy-keyfiles
+		os.MkdirAll(filepath.Join("/run", execUtil.LuksKeyFilePath), 0644)
+		keyFilePath := filepath.Join("/run", execUtil.LuksKeyFilePath, luks.Name)
+		if luks.KeyFile == nil || *luks.KeyFile == "" {
+			// create a keyfile
+			key, err := randHex(4096)
+			if err != nil {
+				return fmt.Errorf("generating keyfile: %v", err)
+			}
+			if err := ioutil.WriteFile(keyFilePath, []byte(key), 0400); err != nil {
+				return fmt.Errorf("creating keyfile: %v", err)
+			}
+			ignitionCreatedKeyFile = true
+		} else {
+			if err := ioutil.WriteFile(keyFilePath, []byte(*luks.KeyFile), 0400); err != nil {
+				return fmt.Errorf("writing keyfile: %v", err)
+			}
+		}
+
+		args := []string{
+			"luksFormat",
+			"--type", "luks2",
+			"--key-file", keyFilePath,
+		}
+
+		if !util.NilOrEmpty(luks.Hash) {
+			args = append(args, "--hash", *luks.Hash)
+		}
+
+		if !util.NilOrEmpty(luks.Label) {
+			args = append(args, "--label", *luks.Label)
+		}
+
+		if !util.NilOrEmpty(luks.UUID) {
+			args = append(args, "--uuid", *luks.UUID)
+		}
+
+		if !util.NilOrEmpty(luks.Cipher) {
+			args = append(args, "--cipher", *luks.Cipher)
+		}
+
+		if len(luks.Options) > 0 {
+			// golang's a really great language...
+			for _, option := range luks.Options {
+				args = append(args, string(option))
+			}
+		}
+
+		args = append(args, luks.Device)
+
+		if _, err := s.Logger.LogCmd(
+			exec.Command(distro.CryptsetupCmd(), args...),
+			"creating %q", luks.Name,
+		); err != nil {
+			return fmt.Errorf("cryptsetup failed: %v", err)
+		}
+
+		// open the device
+		if _, err := s.Logger.LogCmd(
+			exec.Command(distro.CryptsetupCmd(), "luksOpen", luks.Device, luks.Name, "--key-file", keyFilePath),
+			"opening luks device %v", luks.Name,
+		); err != nil {
+			return fmt.Errorf("opening luks device: %v", err)
+		}
+		s.AddCloser(s.closeLuks(luks.Name))
+
+		if luks.Clevis != nil {
+			c := Clevis{
+				Pins: Pin{},
+			}
+			if luks.Clevis.Threshold == nil {
+				c.Threshold = 1
+			} else {
+				c.Threshold = *luks.Clevis.Threshold
+			}
+			for _, tang := range luks.Clevis.Tang {
+				c.Pins.Tang = append(c.Pins.Tang, Tang{
+					URL:        tang.URL,
+					Thumbprint: tang.Thumbprint,
+				})
+			}
+			if luks.Clevis.Tpm2 != nil {
+				c.Pins.Tpm = *luks.Clevis.Tpm2
+			}
+			clevisJson, err := json.Marshal(c)
+			if err != nil {
+				return fmt.Errorf("creating clevis json: %v", err)
+			}
+			if _, err := s.Logger.LogCmd(
+				exec.Command(distro.ClevisCmd(), "luks", "bind", "-f", "-k", keyFilePath, "-d", luks.Device, "sss", string(clevisJson)), "Clevis bind",
+			); err != nil {
+				return fmt.Errorf("binding clevis device: %v", err)
+			}
+		}
+
+		// assume the user does not want a key file, remove it
```

Comment:
Are we sanity-checking anywhere that one of `keyfile` or `clevis` is provided?

Reply:
Yes, the `Validate` function for the `Luks` type will error if `KeyFile` is nil or empty & `Clevis` is nil.